### PR TITLE
Enable `clippy::check-private-items` so that `missing_safety_doc` will apply to private functions as well

### PR DESCRIPTION
**Pull request by 13ros27**
Enabled `check-private-items` in `clippy.toml` and then fixed the resulting errors. Most of these were simply misformatted and of the remaining:
- ~Added `#[allow(clippy::missing_safety_doc)]` to~ Removed unsafe from a pair of functions in `bevy_utils/futures` which are only unsafe so that they can be passed to a function which requires `unsafe fn`
- Removed `unsafe` from `UnsafeWorldCell::observers` as from what I can tell it is always safe like `components`, `bundles` etc. (this should be checked)
- Added safety docs to:
   - `Bundles::get_storage_unchecked`: Based on the function that writes to `dynamic_component_storages`
   - `Bundles::get_storages_unchecked`: Based on the function that writes to `dynamic_bundle_storages`
   - `QueryIterationCursor::init_empty`: Duplicated from `init`
   - `QueryIterationCursor::peek_last`: Thanks Giooschi (also added internal unsafe blocks)
   - `tests::drop_ptr`: Moved safety comment out to the doc string
 
This lint would also apply to `missing_errors_doc`, `missing_panics_doc` and `unnecessary_safety_doc` if we chose to enable any of those at some point, although there is an open [issue](https://github.com/rust-lang/rust-clippy/issues/13074) to separate these options.

### CODE REVIEW

**Comment by 13ros27** [comment]
I don't believe this is a breaking change? The only signature change was `UnsafeWorldCell::observers` which is currently `pub(crate)`

**Review by chescock — APPROVED** [review]

File: crates/bevy_utils/src/futures.rs [hunk 1]
```diff
@@ -36,9 +36,11 @@ pub fn check_ready<F: Future + Unpin>(future: &mut F) -> Option<F::Output> {
     }
 }
 
+#[allow(clippy::missing_safety_doc)] // Unsafe purely due to RawWakerVTable requirements
```
[the remark this answers is not in the document]
Comment (13ros27) [reply crates/bevy_utils/src/futures.rs]: Huh, I didn't know coercion supported that, that's much neater, thanks

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@13ros27 Sorry for the delay on reviewing this; please ping me once merge conflicts are resolved and I'll merge this in.

**Comment by 13ros27** [comment]
No worries, merge conflicts resolved and I've added safety comments to `EntityRefExcept` and `EntityMutExcept` @alice-i-cecile

**Pull request merged**